community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/ConcurrentMapState.java

```diff
@@ -0,0 +1,279 @@
+                    return;
+                }
+            }
+        }
+        synchronized ( value )
+        {
+            update.update( new BigEndianByteArrayBuffer( value ) );
+        }
+    }
+
+    private static class PreviousValue<Key> extends KeyFormat.Searcher<Key> implements KeyValueVisitor{
+        private final byte[] proposal;
+
+        PreviousValue( KeyFormat<Key> keys, Key key, byte[] proposal )
+        {
+            super( keys, key );
+            this.proposal = proposal;
+        }
+
+        @Override
+        public boolean visit( ReadableBuffer key, ReadableBuffer value )
+        {
+            value.get( 0, proposal );
+            return false;
+        }
+    }
+
+    @Override
+    public KeyValueStoreState<Key, Meta> rotate( Meta metadata ) throws IOException
+    {
+        try
+        {
+            Pair<File, KeyValueStoreFile<Meta>> next = rotation.next( file, metadata, keys.filter( dataProvider() ) );
+            return new ConcurrentMapState<>( rotation, keys, next.other(), next.first() );
+        }
+        finally
+        {
+            store.close();
+        }
+    }
+
+    public Meta metadata()
+    {
+        return store.metadata();
+    }
+
+    public KeyValueStoreState<Key, Meta> close() throws IOException
+    {
+        store.close();
+        return null;
+    }
+
+    public boolean hasChanges()
+    {
+        return !changes.isEmpty();
+    }
+
+    public <Value> Value lookup( Key key, AbstractKeyValueStore.Reader<Value> valueReader ) throws IOException
+    {
+        byte[] value = changes.get( key );
+        if ( value != null )
+        {
+            return valueReader.parseValue( new BigEndianByteArrayBuffer( value ) );
+        }
+        ValueFetcher<Key, Value> lookup = new ValueFetcher<>( keys, key, valueReader );
+        if ( store.scan( lookup, lookup ) )
+        {
+            return lookup.value;
+        }
+        return valueReader.defaultValue();
+    }
+
+    private byte[][] sortedUpdates()
+    {
+        Entry[] buffer = new Entry[changes.size()];
+        Iterator<Map.Entry<Key, byte[]>> entries = changes.entrySet().iterator();
+        for ( int i = 0; i < buffer.length; i++ )
+        {
+            Map.Entry<Key, byte[]> next = entries.next(); // we hold the lock, so this should succeed
+            byte[] key = new byte[keys.keySize()];
+            keys.writeKey( next.getKey(), new BigEndianByteArrayBuffer( key ) );
+            buffer[i] = new Entry( key, next.getValue() );
+        }
+        Arrays.sort( buffer );
+        assert !entries.hasNext() : "We hold the lock, so we should see 'size' entries.";
+        byte[][] result = new byte[buffer.length * 2][];
+        for ( int i = 0; i < buffer.length; i++ )
+        {
+            result[i * 2] = buffer[i].key;
+            result[i * 2 + 1] = buffer[i].value;
+        }
+        return result;
+    }
+
+    public DataProvider dataProvider() throws IOException
```

Comment:
Maybe we should add java doc saying that we expect this method to be called only when we hold the lock?